drivers/console/uart_mux.c

```diff
@@ -0,0 +1,821 @@
+			    void *user_data)
+{
+	struct uart_mux_dev_data *dev_data = user_data;
+
+	if (connected) {
+		dev_data->status = UART_MUX_CONNECTED;
+	} else {
+		dev_data->status = UART_MUX_DISCONNECTED;
+	}
+
+	LOG_DBG("%s %s", dev_data->dev->config->name,
+		dev_data->status == UART_MUX_CONNECTED ? "connected" :
+							 "disconnected");
+
+	if (dev_data->attach_cb) {
+		dev_data->attach_cb(dev_data->dev,
+				    dlci ? gsm_dlci_id(dlci) : -1,
+				    connected,
+				    dev_data->attach_user_data);
+	}
+}
+
+static int init_real_uart(struct device *mux, struct device *uart,
+			  struct uart_mux **mux_uart)
+{
+	bool found = false;
+	struct uart_mux *real_uart;
+
+	for (real_uart = __uart_mux_start; real_uart != __uart_mux_end;
+	     real_uart++) {
+		if (real_uart->uart == uart) {
+			found = true;
+			break;
+		}
+	}
+
+	if (found == false) {
+		for (real_uart = __uart_mux_start; real_uart != __uart_mux_end;
+		     real_uart++) {
+			if (real_uart->uart == NULL) {
+				real_uart->uart = uart;
+				found = true;
+				break;
+			}
+		}
+
+		if (found == false) {
+			return -ENOENT;
+		}
+	}
+
+	/* Init the real UART only once */
+	if (atomic_cas(&real_uart->init_done, false, true)) {
+		real_uart->mux = gsm_mux_create(mux);
+
+		LOG_DBG("Initializing UART %s and GSM mux %p",
+			real_uart->uart->config->name, real_uart->mux);
+
+		if (real_uart->mux == false) {
+			real_uart->uart = NULL;
+			return -ENOMEM;
+		}
+
+		uart_irq_rx_disable(real_uart->uart);
+		uart_irq_tx_disable(real_uart->uart);
+		uart_mux_flush_isr(real_uart->uart);
+		uart_irq_callback_user_data_set(
+			real_uart->uart, uart_mux_isr,
+			real_uart);
+
+		k_work_init(&real_uart->rx_work, uart_mux_rx_work);
+		k_mutex_init(&real_uart->lock);
+
+		uart_irq_rx_enable(real_uart->uart);
+	}
+
+	__ASSERT(real_uart->uart, "Real UART not set");
+
+	*mux_uart = real_uart;
+
+	return 0;
+}
+
+/* This will bind the physical (real) UART to this muxed UART */
+static int attach(struct device *mux_uart, struct device *uart,
+		  int dlci_address, uart_mux_attach_cb_t cb,
+		  void *user_data)
+{
+	sys_snode_t *sn, *sns;
+
+	if (mux_uart == NULL || uart == NULL) {
+		return -EINVAL;
+	}
+
+	LOG_DBG("Attach DLCI %d (%s) to %s", dlci_address,
+		mux_uart->config->name, uart->config->name);
+
+	SYS_SLIST_FOR_EACH_NODE_SAFE(&uart_mux_data_devlist, sn, sns) {
+		struct uart_mux_dev_data *dev_data =
+			CONTAINER_OF(sn, struct uart_mux_dev_data, node);
+
+		if (dev_data->dev == mux_uart) {
+			struct uart_mux *real_uart;
+			int ret;
+
+			ret = init_real_uart(mux_uart, uart, &real_uart);
+			if (ret < 0) {
+				return ret;
+			}
+
+			dev_data->real_uart = real_uart;
+			dev_data->status = UART_MUX_CONFIGURED;
+			dev_data->tx_ready = true;
+			dev_data->tx_enabled = true;
+			dev_data->rx_enabled = true;
+			dev_data->attach_cb = cb;
+			dev_data->attach_user_data = user_data;
+
+			dev_data->dlci = gsm_dlci_create(real_uart->mux,
```

Comment:
gsm_dlci_create() failure is not checked here. Perhaps do the check before filling in all the other fields and return error if not created?

Reply:
Yes, good point, will fix.

Reply:
I added checks for the return code, unfortunately the other fields are needed and must be set before the function is called as the `gsm_dlci_create()` will return immediately and those fields are used when negotiating the control channel, otherwise we have null ptr access.